File: forge/templates/python-service-template/template/src/app/middleware/rate_limit.py
"""In-memory per-tenant/per-IP token-bucket rate limiter."""

from __future__ import annotations

import logging
import time
from collections import defaultdict
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field

from fastapi import Request, Response, status
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware

logger = logging.getLogger(__name__)


@dataclass
class _Bucket:
    tokens: float
    last_refill: float = field(default_factory=time.monotonic)


class RateLimitMiddleware(BaseHTTPMiddleware):
    def __init__(
        self,
        app,
        *,
        requests_per_minute: int = 120,
        burst: int | None = None,
        skip_paths: list[str] | None = None,
    ):
        super().__init__(app)
        self._rate = requests_per_minute / 60.0
        self._capacity = float(burst or requests_per_minute)
        self._buckets: dict[str, _Bucket] = defaultdict(lambda: _Bucket(tokens=self._capacity))
        self._skip_paths = set(skip_paths or [])

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        if any(request.url.path.startswith(p) for p in self._skip_paths):
            return await call_next(request)

        key = self._resolve_key(request)
        bucket = self._buckets[key]
        now = time.monotonic()

        elapsed = now - bucket.last_refill
        bucket.tokens = min(self._capacity, bucket.tokens + elapsed * self._rate)
        bucket.last_refill = now

        if bucket.tokens < 1.0:
            retry_after = int((1.0 - bucket.tokens) / self._rate) + 1
            logger.warning("Rate limit exceeded for key=%s", key)
            return JSONResponse(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                content={"detail": "Rate limit exceeded. Please slow down."},
                headers={"Retry-After": str(retry_after)},
            )

        bucket.tokens -= 1.0
        return await call_next(request)

    @staticmethod
    def _resolve_key(request: Request) -> str:
        user = getattr(request.state, "user", None)
        if user is not None:
            customer_id = getattr(user, "customer_id", None)
            if customer_id:
                return f"tenant:{customer_id}"
        if request.client:
            return f"ip:{request.client.host}"
        return "anonymous"

File: forge/features/middleware/templates/rate_limit/rust/files/src/middleware/rate_limit.rs
//! Per-IP token-bucket rate limiter.
//!
//! In-memory only — fine for a single-instance deployment, unsuitable for
//! a horizontally-scaled stack (each replica maintains its own buckets).
//! Swap for a Redis-backed implementation if you deploy multiple replicas.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use axum::body::Body;
use axum::extract::ConnectInfo;
use axum::http::{Request, StatusCode};
use axum::middleware::Next;
use axum::response::{IntoResponse, Response};
use std::net::SocketAddr;

const REQUESTS_PER_MINUTE: f64 = 120.0;
const BURST: f64 = 120.0;

#[derive(Clone)]
pub struct RateLimiter {
    buckets: Arc<Mutex<HashMap<String, Bucket>>>,
    rate_per_sec: f64,
    capacity: f64,
}

struct Bucket {
    tokens: f64,
    last_refill: Instant,
}

impl RateLimiter {
    pub fn new() -> Self {
        Self {
            buckets: Arc::new(Mutex::new(HashMap::new())),
            rate_per_sec: REQUESTS_PER_MINUTE / 60.0,
            capacity: BURST,
        }
    }

    fn check(&self, key: &str) -> bool {
        let mut buckets = self.buckets.lock().unwrap();
        let bucket = buckets.entry(key.to_string()).or_insert_with(|| Bucket {
            tokens: self.capacity,
            last_refill: Instant::now(),
        });
        let now = Instant::now();
        let elapsed = now.duration_since(bucket.last_refill).as_secs_f64();
        bucket.tokens = (bucket.tokens + elapsed * self.rate_per_sec).min(self.capacity);
        bucket.last_refill = now;
        if bucket.tokens < 1.0 {
            return false;
        }
        bucket.tokens -= 1.0;
        true
    }
}

impl Default for RateLimiter {
    fn default() -> Self {
        Self::new()
    }
}

pub async fn rate_limit_middleware(req: Request<Body>, next: Next) -> Response {
    // Skip health/metrics paths so probes aren't rate-limited.
    let path = req.uri().path();
    if path.starts_with("/health") || path.starts_with("/metrics") {
        return next.run(req).await;
    }

    // Pull the client IP from the ConnectInfo extension (set by axum::serve with_connect_info)
    // or fall back to "anonymous" if unavailable (e.g., inside a test harness).
    let key = req
        .extensions()
        .get::<ConnectInfo<SocketAddr>>()
        .map(|ConnectInfo(addr)| addr.ip().to_string())
        .unwrap_or_else(|| "anonymous".to_string());

    // Use a per-process singleton limiter so state persists across requests.
    static LIMITER: std::sync::OnceLock<RateLimiter> = std::sync::OnceLock::new();
    let limiter = LIMITER.get_or_init(RateLimiter::new);

    if !limiter.check(&key) {
        return (
            StatusCode::TOO_MANY_REQUESTS,
            [("retry-after", "60")],
            "Rate limit exceeded. Please slow down.",
        )
            .into_response();
    }
    next.run(req).await
}
